### ISSUE TITLE
mock s3 tests

### ISSUE THREAD

**Issue by mapsam** [mapbox/node-pre-gyp#373]
The tests that use s3 to put/update/delete objects are prone to race conditions if the same tests are being run from multiple places (i.e. pr & push builds on travis). This leads to unexpectedly flopping tests and solving with a retry, which is a bit cumbersome.

In order to run tests reliably, we can mock out s3 functions and their responses and assert specific parameters were called. Let's have a look at [mock-aws-sdk-js](https://github.com/mapbox/mock-aws-sdk-js), which uses sinon stubs.

cc @springmeyer

**Comment by springmeyer** [comment]
> Let's have a look at mock-aws-sdk-js, which uses sinon stubs.

Or https://github.com/node-nock/nock perhaps as I think I saw @agius using this?

**Comment by mapsam** [comment]
We can use nock for direct HTTP requests, whereas sinon/mock-aws-sdk-js would allow us to mock the aws-sdk methods like `s3.getObject`. Do we use both for s3 interaction?

**Comment by springmeyer** [comment]
Ah, okay. In this case node-pre-gyp primarily uses both http and aws-sdk to interact with AWS. http for GET and aws-sdk for PUT/DELETE/LIST.

**Comment by agius** [comment]
Nock was actually very difficult to use to mock aws-sdk. Simple things like s3 GET requests can cascade into multiple requests, and it took tons of effort to figure out what kind of response they all wanted. I have given up that approach in favor of object mocking libraries ( [sinon](http://sinonjs.org/releases/v5.0.6/) , [mock-aws-sdk-js](https://github.com/mapbox/mock-aws-sdk-js) , [testdouble](https://github.com/testdouble/testdouble.js) , or [jest](https://facebook.github.io/jest/) )

**Comment by mapsam** [comment]
Agreed @agius - I've always used mock-aws-sdk-js (sinon basically) for aws-sdk, and nock for pure HTTP requests. More recently I've just been using sinon to mock request.js though - with a `sinon.stub(request, 'get')` and that seems to suffice!

**Comment by mapsam** [comment]
Hmm, running into a roadblock trying to mock s3 methods that are called within a Node.js child process, which comes from `test/run.test.js` and all of the app/build tests. If we want to mock these s3 calls, we'll need to mock within the proper process ID, or rewrite the test/build.test.js tests to call actual functions rather than the bin commands, which doesn't sound half bad.

**Comment by mapsam** [comment]
Diving into these tests has left me with some large-scale refactor desires. Opened a ticket in https://github.com/mapbox/node-pre-gyp/issues/392 to discuss this further.
